kotishell/src/tokenize/tokenize_a.c

```diff
@@ -6,39 +6,42 @@
 /*   By: elenavoronin <elnvoronin@gmail.com>          +#+                     */
 /*                                                   +#+                      */
 /*   Created: 2023/06/29 16:20:23 by elenavoroni   #+#    #+#                 */
-/*   Updated: 2023/07/13 18:54:20 by dkolodze      ########   odam.nl         */
+/*   Updated: 2023/07/14 15:59:05 by evoronin      ########   odam.nl         */
 /*                                                                            */
 /* ************************************************************************** */
 
 #include "tokenize.h"
 
 static void	s_tk_whitespace(t_tk_so_far *so_far, char *s, t_tk_result *result)
 {
-	if (so_far->token.type != TK_WHITESPACE && so_far->token.length == 0)
+	if (so_far->token.type == TK_EMPTY)
 	{
 		so_far->token.length = 1;
 		so_far->token.type = TK_WHITESPACE;
 		so_far->token.data = s;
 	}
 	else
 		so_far->token.length += 1;
-	if (tk_next_state(so_far->state, s + 1) != so_far->state)
+	if (s_tk_next_state(so_far->state, s + 1) != so_far->state)
 	{
-		tk_token_copy(so_far, result);
+		s_tk_token_copy(so_far, result);
+		so_far->token.type = TK_EMPTY;
 		so_far->token.length = 0;
-		so_far->state = tk_next_state(so_far->state, s + 1);
+		so_far->state = s_tk_next_state(so_far->state, s + 1);
 	}
 }
 
 static void	s_tk_end(t_tk_so_far *so_far, char *s, t_tk_result *result)
 {
-	if (so_far->token.type != TK_EOL)
+	if (so_far->token.type == TK_EMPTY)
 	{
 		so_far->token.type = TK_EOL;
 		so_far->token.length = 0;
 		so_far->token.data = s;
 	}
-	tk_token_copy(so_far, result);
+	s_tk_token_copy(so_far, result);
+	if (so_far->status == TK_ERR_MALLOC)
+		s_tk_error(so_far, result);
```

Comment:
Let's call s_tk_error once only in tk_end; and add guard (check if the status is not TK_SUCCESS) that will prevent all the functions from doing things